### Scoring using the best model locally - Optional

The training environment in AML compute cluster can be different than the one locally. We need to create a new Conda environment in the repo so that model can be loaded locally. 

    conda env create -f conda_env.yml


Attach this notebook to the new conda environment. 

Use the model file downloaded using automl_amlsdkv2_training.ipynb. 

If AML Compute Instance is used, then the notebook kernel should be set to Python 3.8 - AzureML

Load the model

In [ ]:
import joblib
import os

model_path = os.path.join(".././artifact_downloads/outputs", "mlflow-model" ,"model.pkl")
model = joblib.load(model_path)

In [7]:
model

PipelineWithYTransformations(Pipeline={'memory': None,
                                       'steps': [('datatransformer',
                                                  DataTransformer(enable_dnn=False, enable_feature_sweeping=False, feature_sweeping_config={}, feature_sweeping_timeout=86400, featurization_config=None, force_text_dnn=False, is_cross_validation=True, is_onnx_compatible=True, observer=None, task='classification', working_dir='c:\...
                                                  PreFittedSoftVotingClassifier(classification_labels=array([0, 1], dtype=int64), estimators=[('1', Pipeline(memory=None, steps=[('maxabsscaler', MaxAbsScaler(copy=True)), ('lightgbmclassifier', LightGBMClassifier(boosting_type='gbdt', colsample_bytree=0.6933333333333332, learning_rate=0.09473736842105263, max_bin=110, max_depth=8, min_child_weight=6, min_data_in_leaf=0.003457931034482759, min_split_gain=1, n_estimators=25, n_jobs=-1, num_leaves=227, problem_info=ProblemInfo(gpu_training_pa

Use the test dataset created in automl_amlsdkv2_training.ipynb

Format the test dataset and drop the Churn column. 

In [2]:
my_unique_dataset_name = "1155c2a891714097b099ade1fee4bddb" # Replace with your unique dataset name
my_unique_test_data_set_path = f'.././telcocustomerchurn/{my_unique_dataset_name}_test/{my_unique_dataset_name}-WA_Fn-UseC_-Telco-Customer-Churn_Test.csv'

In [4]:
import pandas as pd

test_data_orig = pd.read_csv(my_unique_test_data_set_path)

test_data = test_data_orig.drop("Churn", axis=1)
test_data["Partner"] = test_data["Partner"].map({"Yes": True, "No": False})
test_data["Dependents"] = test_data["Dependents"].map({"Yes": True, "No": False})
test_data["PhoneService"] = test_data["PhoneService"].map({"Yes": True, "No": False})
test_data["PaperlessBilling"] = test_data["PaperlessBilling"].map({"Yes": True, "No": False})
 


Run the prediction

In [8]:

result = model.predict(test_data)

Merge prediction results with test data to compare the results. 

In [9]:
test_data_orig["Churn_Prediction"] = result

test_data_orig['Churn_Prediction'] = test_data_orig['Churn_Prediction'].map({True: 'Yes', False: 'No'})


In [10]:
test_data_orig.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn',
       'Churn_Prediction'],
      dtype='object')

In [11]:
selected_columns = test_data_orig[['customerID', 'Churn', 'Churn_Prediction']]

# Display the selected columns
print(selected_columns)

      customerID Churn Churn_Prediction
0     4376-KFVRS    No               No
1     2754-SDJRD    No              Yes
2     9917-KWRBE    No               No
3     0365-GXEZS    No               No
4     9385-NXKDA    No               No
...          ...   ...              ...
1404  5204-HMGYF    No               No
1405  9950-MTGYX    No               No
1406  3675-EQOZA    No               No
1407  3646-ITDGM    No               No
1408  3913-FCUUW    No               No

[1409 rows x 3 columns]


Calculate accuracy, precision, recall, and F1-score

In [12]:
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Assuming you have a DataFrame called 'test_data_orig'
# with 'Churn' representing actual labels ('No' or 'Yes') and 'Churn_Prediction' representing predicted labels ('No' or 'Yes')
actual_labels = (test_data_orig['Churn'] == 'Yes').astype(int)  # Convert 'Yes' to 1 and 'No' to 0
predicted_labels = (test_data_orig['Churn_Prediction'] == 'Yes').astype(int)  # Convert 'Yes' to 1 and 'No' to 0

# Calculate accuracy, precision, recall, and F1-score
accuracy = accuracy_score(actual_labels, predicted_labels)
precision = precision_score(actual_labels, predicted_labels)
recall = recall_score(actual_labels, predicted_labels)
f1 = f1_score(actual_labels, predicted_labels)

# Print the metrics
print(f"Accuracy: {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall: {recall:.2f}")
print(f"F1-Score: {f1:.2f}")


Accuracy: 0.80
Precision: 0.65
Recall: 0.51
F1-Score: 0.57


Accuracy (0.80): An accuracy of 80% indicates that your model correctly predicts whether a customer will churn or not in 80% of the cases. This is a reasonably good starting point, but it's important to consider other metrics due to the potential class imbalance in customer churn problems. For example, if only a small percentage of customers actually churn, a model that predicts "no churn" for all customers could still achieve a high accuracy, but it would not be useful.

Precision (0.65): Precision of 65% means that when your model predicts a customer will churn, it is correct 65% of the time. This metric is important because you want to avoid incorrectly flagging customers who don't intend to churn as potential churners. A precision of 65% suggests that the model is decent at identifying actual churners without too many false positives.

Recall (0.51): Recall of 51% means that your model is able to capture 51% of the actual churn cases. While this isn't extremely high, it's still meaningful because it indicates that the model can identify more than half of the customers who are likely to churn. Recall is important when you want to minimize false negatives and ensure you don't miss potential churners.

F1-Score (0.57): The F1-Score of 0.57 is a balanced metric that takes both precision and recall into account. It suggests that the model strikes a balance between identifying actual churners and avoiding false alarms.

In customer churn prediction, it's often more important to have a higher recall (to capture as many potential churners as possible) while maintaining a reasonable precision (to avoid unnecessarily alarming non-churners). However, the balance between precision and recall depends on the specific business goals and costs associated with churn.

Consider monitoring these metrics over time, and if possible, compare your model's performance to previous methods or industry benchmarks to assess whether further improvements are needed. Additionally, you might want to perform more advanced analyses, such as feature importance assessment or model tuning, to enhance your churn prediction model.